### ISSUE TITLE
Mechanisms for not 'trusting' docblock types

### ISSUE THREAD

**Issue by ciaranmcnulty** [vimeo/psalm#8017]
This spins out from #8005 and some slack conversations

In general Psalm treats docblock types as truth. This is great for an application because I can add `@param array<Foo>`, no longer feel like I have to validate my array, and any calling code that doesn't obey the type definition is caught.

For libraries this is different - specific APIs are going to be accessed by code that may not be validated using Psalm (or other SA). This leaves the author with choices

  - Decide that callers who ignore the type are 'wrong' and any errors are part of 'undefined behaviour' and their own fault
  - Write redundant checking code that Psalm will complain is unnecessary
  - De-activate the param somehow (e.g. `@var mixed` inside the body of the function)

IMO we should have a way to opt out of this behaviour either globally, per method, or using an internal/external mechanism:

## Global opt-out

Any validated callers would still need to pass the right types, but when evaluating the code inside the function the docblock param would effectively be ignored 

```xml
<psalm
  useDocblockParamTypesInFunctionBody="false"
/>
```
```php
/** @param non-empty-array<Foo> $a */
function f(array $a): Foo
{
   return current($a); // MixedReturnStatement here
}
```

## Per-functions opt-out

It may be we only want to do this for certain functions that we 'know' are going to be called from an untrusted source. In that case a flag in the docblock would make sense:

```php
/** 
 * @param non-empty-array<Foo> $a 
 * @psalm-ignore-docblock-params
 */
function f(array $a): Foo
{
   return current($a); // MixedReturnStatement here
}
```

## 'External' opt-out

Many libraries have an existing concept of what is internal or external to them. We could have a setting that uses that existing labelling

```xml
<psalm
  useDocblockParamTypesInExternalFunctionBody="false"
/>
```
```php
/** 
 * @param non-empty-array<Foo> $a 
 */
function f(array $a): Foo
{
   return current($a); // MixedReturnStatement
}

/** 
 * @param non-empty-array<Foo> $a 
 * @internal
 */
function g(array $a): Foo
{
   return current($a); // no error here because we assume all callers are validated
}
```

**Comment by kkmuffme** [comment]
I think that's a bad idea, as there already is a way to fix this, that is the generally accepted way to do it (and works with phpstorm, phpstan too):
use @var to declare the actual type of the variable.

To avoid this doing over and over again for a specific function from a library, the way to go is, to use stubs:
Just stub your external libraries and add it your repo in a separate (e.g. my-stubs) branch. Merge that branch to your main.
You can fix the types of individual functions in the stubs (or remove the types from the stubs alltogether if they're all wrong) on your main or any other branch.
Everytime the library changes, you can recreate the stubs in that branch (my-stubs) and merge it to your main again, so your changes to the types are not lost, even when the library updates.
This is essentially the git fork model (with fetch upstream) in your repo.

**Comment by ciaranmcnulty** [comment]
@kkmuffme  I'm talking about when writing a library, so don't know how stubs come into it?

**Comment by weirdan** [comment]
> Thanks for writing this up as a clear feature request.

That's implemented already: https://packagist.org/packages/roave/you-are-using-it-wrong

**Comment by Ocramius** [comment]
> specific APIs are going to be accessed by code that may not be validated using Psalm (or other SA). This leaves the author with choices

Two ways forward here, in my opinion:

 1. write adapters for everything, and use `/** @var mixed */ $this->someLibraryDependency->something()`, then `assert()` on everything
 2. trust the third-party to provide valid type annotations. If the library breaks that declaration, that's a breach of trust: trust is how the entire ecosystem works, since SemVer is just a fancy way of saying "I trust you to a certain degree".

My endorsement:

 1. reduce dependencies to a bare minimum (always do this, regardless)
 2. if you must a dependency, use one from a community that you trust, not something random.

As for `roave/you-are-using-it-wrong`, it is mostly aimed at consumers of a library, and forcing them to use the API as documented in its types. This works for **applications**, but **not for libraries**.

Overall, I fully understand the OP's concern, but either we make psalm a pre-deploy step (or similar), or we change to a language where these breaches in the type system are not viable at all. In PHP, it is more a community problem, than a technical one.

**Comment by AndrolGenhald** [comment]
@Ocramius I think this issue is from the perspective of the library author, it sounds like you're writing from the perspective of a library consumer.

As a library author it sounds to me like it might be nice to both

 - Have docblock types that are correct
 - Have Psalm allow/enforce type checks to ensure callers use the correct types

Maybe roave/you-are-using-it-wrong handles this well enough already, I haven't looked at it too much, but Psalm support for this isn't difficult and it seems like it might be useful in addition to (or instead of) roave/you-are-using-it-wrong?

**Comment by Ocramius** [comment]
If this is solely from the library author PoV looking at downstream (attempting to get consumers to respect the types), then just use `roave/you-are-using-it-wrong`: it is very minimal, and mostly just needs some polishing to automate dependency upgrades and associated testing.

It already does what is being asked in OP.

Nothing can be done if `vimeo/psalm` is not used downstream: that's literally a "Free as in read the OSS license" problem for consumers.

In addition to that, changing the type system between internal and external functions is just complicating things further, introducing a further level of "taint checking" around type precision which is not worth pursuing.

Alternatively, declare all your public API inputs as `mixed`, and `assert()` on everything there, then suppress all redundant assertion checks (if the types are documented with precision): that already works out of the box.

**Comment by AndrolGenhald** [comment]
> Alternatively, declare all your public API inputs as mixed, and assert() on everything there, then suppress all redundant assertion checks (if the types are documented with precision): that already works out of the box.

What I was thinking of was basically a docblock tag to do that automatically but without having a bunch of `@var` and/or `@psalm-suppress`.

**Comment by Ocramius** [comment]
That's already achieved by doing ( https://psalm.dev/r/e413cb9354 ):

```php
/** @param array{foo: int} $input */
function foo(mixed $input): void
{
    assert(is_array($input));
    assert(array_key_exists('foo', $input));
    assert(is_int($input['foo']));
}
```


You can now decide to:

 1. declare `/** @var mixed $input */` at the start of your function ( https://psalm.dev/r/900478b038 )
 2. suppress `RedundantConditionGivenDocblockType` for all your public-API-facing files

**Comment by AndrolGenhald** [comment]
True, it's already possible to do with `@var` or `@psalm-suppress`, but personally I view both of those tags as things that should be periodically reviewed to see if they can be improved, because both are an indication to me that something doesn't work the way it ideally should. It also requires having a bunch of extra `@var` or `@psalm-suppress` tags, and while I'd personally prefer `@var`, then you have to keep the `@var` declarations in sync with the PHP param declarations, and any mistakes are impossible for Psalm to catch.

The implementation of a docblock tag for this was trivial, so I just went ahead and did it to show how I think it should work.

**Comment by AndrolGenhald** [comment]
Some additional considerations for this vs `roave/you-are-using-it-wrong`:

 - `you-are-using-it-wrong` is tied to `composer install/update`, so it will only show issues when that happens, whereas this solution will force the library to handle incorrect types likely by throwing an exception immediately, so it will be impossible to miss when the consumer is testing their code
   - `composer install` isn't always run frequently, I've seen manual deployments where it's skipped
   - Even with CD `composer install` might be skipped to save time based on a hash of `composer.lock`
 - `you-are-using-it-wrong` is ignored if the project has a Psalm config, but that Psalm config might be ignoring issues that the library author cares about

Don't get me wrong, `you-are-using-it-wrong` seems pretty great, I'm just not sure its existence is enough reason to avoid implementing this feature.

**Comment by AndrolGenhald** [comment]
Any further thoughts on this? Is anyone actually _opposed_ to it given that the implementation is trivial and not likely to be a maintenance burden?

**Comment by AndrolGenhald** [comment]
> Please add proper docs for it, so it doesn't end up as an undocumented psalm feature like it happened often in the past and people know how to use this.

Yeah, there are some documentation tests and I've been adding more over time to make sure this doesn't happen, you can see there's actually a doc test failing for the PR since I didn't bother writing up documentation at the time.

**Comment by VincentLanglet** [comment]
Phpstan has a `treatPhpdocAsCertain: false` solution.

For psalm, I generally add 
```
    <issueHandlers>
        <!-- Psalm equivalent of PHPStan config `treatPhpDocTypesAsCertain: false` -->
        <DocblockTypeContradiction errorLevel="suppress"/>
        <RedundantConditionGivenDocblockType errorLevel="suppress"/>
    </issueHandlers>
```
In my config, and it does the work.

This tips maybe need to be documented, but IMHO that's enough since that especially the purpose of ignoring those errors.

**Comment by AndrolGenhald** [comment]
> Phpstan has a `treatPhpdocAsCertain: false` solution.

I'm actually a bit surprised how that works: https://phpstan.org/r/fb2239d7-c349-4a2f-bf92-030002822a05
I expected it to show an error there.

This change for Psalm would actually make it much more strict, and would treat `$bar` as `mixed` in that case until you did a `if (!is_string($bar)) throw new InvalidArgumentException()` check.

**Comment by VincentLanglet** [comment]
> > Phpstan has a `treatPhpdocAsCertain: false` solution.
> 
> I'm actually a bit surprised how that works: https://phpstan.org/r/fb2239d7-c349-4a2f-bf92-030002822a05 I expected it to show an error there.

Using `treatPhpdocAsCertain: false` doesn't ignore the phpdoc but allows extra safety check.
So it doesn't force you to always check the params, but it allows you to do it.

This way https://phpstan.org/r/129986bb-418e-434e-89d9-5f7137e44894 doesn't give an error (when https://phpstan.org/r/5fd53dab-ddcb-4949-b399-484d9bbb5efb does).
